File: src/main/java/com/raychenon/leetcode/IntegerToRoman.java
package com.raychenon.leetcode;

import java.util.Arrays;
import java.util.List;

/**
 * User: raychenon
 * Date: 2019-09-12
 * https://leetcode.com/problems/integer-to-roman/
 */
public class IntegerToRoman {

    private List<Tuple> romanList = Arrays.asList(
            new Tuple(1000, "M"),
            new Tuple(900, "CM"),
            new Tuple(500, "D"),
            new Tuple(400, "CD"),
            new Tuple(100, "C"),
            new Tuple(90, "XC"),
            new Tuple(50, "L"),
            new Tuple(40, "IV"),
            new Tuple(10, "X"),
            new Tuple(9, "IX"),
            new Tuple(5, "V"),
            new Tuple(4, "IV"),
            new Tuple(1, "I"));

    static class Tuple {
        Integer integer;
        String roman;

        public Tuple(Integer integer, String roman) {
            this.integer = integer;
            this.roman = roman;
        }
    }

    public String intToRoman(int num) {
        int index = 0;
        StringBuilder str = new StringBuilder();
        while (num != 0) {
            if (num >= romanList.get(index).integer) {
                Tuple tuple = romanList.get(index);
                num = num - tuple.integer;
                str.append(tuple.roman);
            } else {
                index++;
            }
        }
        return str.toString();
    }

}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/IntegerToRoman.kt
package com.raychenon.kotlin.leetcode

/**
 * User: raychenon
 * Date: 2019-09-12
 * https://leetcode.com/problems/integer-to-roman/
 */
object IntegerToRoman {
    
    val romanList = listOf<Pair<Int, String>>(
        Pair(1000, "M"),
        Pair(900, "CM"),
        Pair(500, "D"),
        Pair(400, "CD"),
        Pair(100, "C"),
        Pair(90, "XC"),
        Pair(50, "L"),
        Pair(40, "XL"),
        Pair(10, "X"),
        Pair(9, "IX"),
        Pair(5, "V"),
        Pair(4, "IV"),
        Pair(1, "I")
    )
    
    fun intToRoman(num: Int): String {
        var number: Int = num // Kotlin method parameters are passed as val(immutable)
        var index = 0
        var str = StringBuffer()
        while (number != 0) {

            if (number >= romanList[index].first) {
                val romanValue = romanList[index]
                number = number - romanValue.first
                str.append(romanValue.second)
            } else {
                index = index + 1
            }
        }
        return str.toString()
    }
}